### DISCUSSION TITLE
Just a "fuller" version of ImFontAtlas::GetGlyphRangesChineseFull()

### DISCUSSION THREAD

**Discussion by OKOKDragon** [ocornut/imgui#6693] [New Users Build/Link/Run issues ONLY!]
Well, I am a Chinese user. Although `GetGlyphRangesChineseFull()` can cover a majority of Chinese characters, there are still some symbols which I think every user may need but not loaded by default. For example, I want tabs like `'│'` and `'─'` , and I hate to simply replace them with `'|'` and `'—'`. There are complex Chinese characters like `'﨔'` and `'凞'`, which may not be used in modern daily life, but are commonly seen in ancient Chinese master pieces of literature. So I read the complete utf-8 character list and make a conclusion of everything a Chinese user might need.

```c++
const ImWchar* ImFontAtlas::GetGlyphRangesChineseEverything()
{
    static const ImWchar ranges[] =
    {
        0x0020, 0xAC00, 
        0xD7A3, 0xD7A3,
        0xD7B0,0xD7FB,
        0xE000,0xE000,
        0xF8FF,0xFB06,
        0xFB13,0xFB17,
        0xFB1D,0xFBC2,
        0xFBD3,0xFDCF,
        0xFDF0,0xFDFF,
        0xFE10,0xFEFC,
        0xFEFF,0xFEFF,
        0xFF01,0xFFFD,
        0,
    };
    return &ranges[0];
}
``` 
This is a long list and, of course, makes your program much slower when it starts. There are ways to make it faster to load your fonts, and I recommend this:
write every character you need in a `.txt` file, and add following codes.
```c++
ImGuiIO& io = ImGui::GetIO(); (void)io;

FILE* inFile = fopen("Asset/Font/chinese7000.txt", "rb");

unsigned char *charBuf;

fseek(inFile, 0, SEEK_END);
int fileLen = ftell(inFile);
charBuf = new unsigned char[fileLen];
fseek(inFile, 0, SEEK_SET);
fread(charBuf, fileLen, 1, inFile);

fclose(inFile);

static ImVector<ImWchar> myRange;
ImFontGlyphRangesBuilder myGlyph;


myGlyph.AddText((const char*)charBuf);
myGlyph.BuildRanges(&myRange);

delete[] charBuf;


ImFont* font1 = io.Fonts->AddFontFromFileTTF("Asset/Font/SourceHanSansCN-Normal.otf", 14.0f, nullptr, myRange.Data);
ImFont* font2 = io.Fonts->AddFontFromFileTTF("Asset/Font/Alibaba-PuHuiTi-Bold.ttf", 18.0f, nullptr, myRange.Data);
``` 

This method is originally written by "[bianruifeng](https://www.jianshu.com/u/46965e5a5292)". Read [the original article](https://www.jianshu.com/p/dd029e6d9dd4) for more details.

Besides, for Chinese users new to fonts and characters in computers, mind this:
never use fonts whose name contains "GB", which refers to "国标". For example, GB2312 defines only 6763 very common characters, and you will never complete an  article with any terminology using them.

**Comment by tomasmark79** [comment]
This topic helped me safe some time. Lot of hours spent with why my Czech language chars are represented like question marks.
